libs/nio/src/main/java/org/elasticsearch/nio/NioSocketChannel.java

```diff
@@ -27,15 +27,20 @@
 
 public class NioSocketChannel extends NioChannel {
 
-    private final InetSocketAddress remoteAddress;
     private final AtomicBoolean contextSet = new AtomicBoolean(false);
     private final SocketChannel socketChannel;
+    private volatile InetSocketAddress remoteAddress;
+    private volatile InetSocketAddress localAddress;
     private SocketChannelContext context;
 
-    public NioSocketChannel(SocketChannel socketChannel) throws IOException {
-        super(socketChannel);
+    public NioSocketChannel(SocketChannel socketChannel) {
         this.socketChannel = socketChannel;
-        this.remoteAddress = (InetSocketAddress) socketChannel.getRemoteAddress();
+        try {
+            remoteAddress = (InetSocketAddress) socketChannel.getRemoteAddress();
+        } catch (IOException e) {
+            // This exception will be thrown if the channel is closed. But we do not really care when
+            // we are just attempting to read the remote address.
```

Comment:
can we somehow assert that it's closed?

Reply:
I think that is dangerous because that is not guaranteed in the method contract. It is just from me reading the `SocketChannelImpl.java`. I changed it to throw an UncheckedIOException (which is caught in the `ChannelFactory`)